### ISSUE TITLE
Have syscheck ignore events on directories excluded by recursion level

### ISSUE THREAD

**Issue by Molter73** [wazuh/wazuh#5249]
|Wazuh version|Component|Platform|
|---|---|---|
| 3.13 | syscheckd | Linux/Windows |

## Description

Currently syscheck is not using the `recursion_level` set on directories configuration to ignore events from realtime and whodata engines or stop scanning a directory on scheduled mode, this causes a waste of processing resources on agents.

### Scheduled mode
When finding a directory nested below the set `recursion_level`, said directory should be ignored by the scan and move on. In other words, when reaching the set `recursion_level` on a directory, files should be scanned and nested directories ignored.

#### Linux
Watches should not be added to directories beyond the configured `recursion_level`, this will save up processing power on both the agent and the inotify engine.

#### Windows
Instead of watching directories with a recursive callback, it should be set on each individual directory up to the configured `recursion_level`. If no `recursion_level` is configured, setting a recursive watch might be more efficient.

#### Linux
A mechanism should be put in place so that audit rules for ignoring directories beyond the configured `recursion_level` are added, this will save up processing power on the agent, but care should be taken since auditd will have more rules to process and might end up consuming more resources.

#### Windows
The audit policies should be added to directories up to the configured `recursion_level`.

**Comment by Molter73** [comment]
After [this PR](https://github.com/wazuh/wazuh/pull/5144) was merged, scans on any directory are stopped once the recursion level is reached, whereas previously the files and directories that were inside a directory on the set recursion level were grabbed and discarded. This also prevents inotify watches from being added to directories deeper than the configured recursion level.

Going over how the audit system processes events, adding rules meant for ignoring events on directories deeper than the configured recursion level would not only incur extra processing on the kernel side for unrelated operations, it could also potentially interfere with other rules set by users. We should probably disregard this part of the issue.

On the Windows side, using individual handles for monitoring in realtime instead of individual recursive ones would increase the amount of memory used, since a buffer associated with each handle is reserved for events (as described in the [remarks section of the MSDN entry for ReadDirectoryChangesW](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-readdirectorychangesw#remarks)). This combined with the effort required to change the code to properly work with non recursive watches make this change quite less appealing.

Finally, with Windows ACEs and SACLs, it looks like by default the ACEs we are setting propagate to child objects by default, it could be worth trying to add some of the inheritance flags described in the [ACE_HEADER MSDN entry](https://docs.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-ace_header#members) and see if we can prevent our ACE from being inherited to directories beyond the configured recursion level.

**Comment by Molter73** [comment]
After attempting to modify the inheritance values configured by Wazuh as described in my previous comment, I've found that there are ways in which a python script (as an example) could create a series of nested directories and trigger events that only have `SYNCHRONIZE` and `ListDirectory` in its permission mask. These 2 properties are not only common but the same ones that Wazuh uses when it gets information from a directory (such as listing its contents for scanning), which ends up creating an infinite loop of triggering events while trying to process those same events.

I've tried different configurations of inheritance and permissions, but in the end it seems that the only way around these types of events is to use inheritable permissions as the ones we already have.

I have double checked that all possible modes of operations on all supported OSes stop processing as soon as possible if the recursion level is exceeded, so it seems like no further improvements are possible on this subject.
